### ISSUE TITLE
Track selection is wrong when using watch-later

### ISSUE THREAD

**Issue by po5** [CogentRedTester/mpv-coverart#2]
Repro steps:
- open audio file with external cover art
- run `quit-watch-later`
- open the file again

Instead of loading the external cover, the script will act as if `--vid=no` was passed.
What happens is that `vid=1` is saved to the watch_later file, as that's the currently selected video track. On next load, it will fall back to `vid=no` because there is no video track.
I don't know of a way to differentiate between this behavior and `--vid=no` + `--start`.

A sane solution would be to add an option that always forces the first cover to be selected, and use regular logic for the others.
Such an option would not play nice with `always_scan_coverart` if playing a directory with a bunch images including some that match our cover art requirements, but that's life.

**Comment by CogentRedTester** [comment]
Are you sure? This doesn't happen for me, the external cover art loads as expected. Are you using any non-default settings?

**Comment by po5** [comment]
I was able to reproduce this with the latest coverart.lua from the repo and nothing else in my mpv config dir.

**Comment by CogentRedTester** [comment]
Well I can't reproduce it, I'm running mpv with the following settings:

`mpv --msg-level=coverart=v 'file.flac' --save-position-on-quit --watch-later-  directory=C:\Users\--\AppData\Roaming\mpv\portable_config\watch_later --no-config --script=coverart.lua`

```
mpv 0.32.0-563-gdc24a437fb Copyright © 2000-2020 mpv/MPlayer/mplayer2 projects
 built on Sun Jun 21 22:18:57 +08 2020
FFmpeg library versions:
   libavutil       56.55.100
   libavcodec      58.93.100
   libavformat     58.47.100
   libswscale      5.8.100
   libavfilter     7.86.100
   libswresample   3.8.100
FFmpeg version: git-2020-06-21-842bc312a
```

**Comment by po5** [comment]
Can reproduce with that exact command, on commit https://github.com/mpv-player/mpv/commit/dc24a437fb9ba2485bc8c5866542bd36254410d4 (same one as yours if using shinchiro builds on windows)

To be clear, you have to run the command twice on a file with no embedded cover art and a cover.jpg or other present in the same folder to see the effect.
I doubt that this could be an OS issue as the `vid` fallback is done by mpv but I'm out of ideas.

**Comment by CogentRedTester** [comment]
Can you upload a log?

**Comment by CogentRedTester** [comment]
It may be worth noting there are two vid properties at work here; `vid` which tracks the currently selected track, and `options/vid` which tracks the option passed to --vid when mpv starts.

coverart uses `options/vid` to determine if it needs to auto-load the file, and I believe that this is what is saved in the watch_later directory, regardless of what the current video actually is. So if you launch mpv with auto that seems to be what is sent to the script when reopening the file. If you launch with --vid=no and change to another track, then it'll go back to --vid=no when reloading the file. That's what my experimentation has uncovered.

Presumable mpv has some internals which change the track to the actually previously selected one, but it doesn't work on external files added after playback starts, so we can't do anything about that.

**Comment by po5** [comment]
Logs: https://gist.github.com/po5/1e8a4db81096ac0bdc1a9b5825dd0394

As you can see on first run (empty watch_later dir)
`[cplayer] Run command: video-add, flags=64, args=["/run/media/eva/Music/rdm/(2012) VA - SKETCH UP! REMIXIES [16FLAC]/folder.jpg", "select", "", ""]`
then on 2nd run
`[cplayer] Run command: video-add, flags=64, args=["/run/media/eva/Music/rdm/(2012) VA - SKETCH UP! REMIXIES [16FLAC]/folder.jpg", "auto", "", ""]`

https://github.com/CogentRedTester/mpv-coverart/blob/3b3c82253b19c82ae5965baf6c69f4a81d6783fb/coverart.lua#L201-L205 Where `mp.get_property('vid')` returns "no", and because it's not a number `mp.get_property_number('vid', 0)` returns the fallback value of 0.
`options/vid` however returns 1 which isn't auto and won't pass the check.

I may have gotten confused explaining this initially because of these two properties.

**Comment by CogentRedTester** [comment]
Hmm, very interesting. I'll take another look when I get home and see how my computer differs. It does seem like something is going on here.

**Comment by CogentRedTester** [comment]
Ok, my config file is as expected, 'select' is being run both times, never auto. I'm still going through our logs, but it seems you're using an older version of mpv player. You're using the 0.32 release, I'm on a newer git release from the last couple of weeks, the behaviour of watch_later probably changed during that time.

**Comment by CogentRedTester** [comment]
Yes, I knew I remembered something like this happening, see the note in the latest [manual](https://mpv.io/manual/master/#options-aid).

This is the relevant section:

> At mpv 0.32.0 (and some releases before), if you passed a track value for which a corresponding track didn't exist (e.g. --aid=2 and there was only 1 audio track), the aid property returned no. However if another audio track was added during playback, and you tried to set the aid property to 2, nothing happened, because the aid option still had the value 2, and writing the same value has no effect.
> 
> With mpv 0.33.0, the behavior was changed. Now track selection options are reset to auto at playback initialization, if the option had tries to select a track that does not exist. The same is done if the track exists, but fails to initialize. The consequence is that unlike before mpv 0.33.0, the user's track selection parameters are clobbered in certain situations.
> 
> Also since mpv 0.33.0, trying to select a track by number will strictly select this track. Before this change, trying to select a track which did not exist would fall back to track default selection at playback initialization. The new behavior is more consistent.
> 
>

**Comment by po5** [comment]
Thank you so much for noticing that.
Turns out I had installed mpv instead of mpv-git when reinstalling my system.
Works on 0.33, sorry for wasting your time.

**Comment by CogentRedTester** [comment]
Using the new preload option now fixes this on versions prior to 0.33 as well.
